### PR TITLE
Functions about syndrome decoding

### PR DESCRIPTION
**Pull request by RexWzh**
# Main change
New functions: `polynomial_eval, syndrome_polynomial, haserrors, erratalocator_polynomial, error_evaluator`

About Manifest.toml:
   - generated by `]add https://github.com/JuliaImages/QRCoders.jl#dev`
   - manually edit `uuid, version` of `QRCoders`
     ```toml
     [[QRCoders]]
     deps = ["FileIO", "ImageCore", "ImageIO"]
     git-tree-sha1 = "07504090c0bf95217cadde745316aeaacbf7f363"
     repo-rev = "dev"
     repo-url = "https://github.com/JuliaImages/QRCoders.jl"
     uuid = "f42e9828-16f3-11ed-2883-9126170b272d"
     version = "0.1.0"
     ```

### CODE REVIEW

**Comment by RexWzh** [comment]
## Implementation plan
Let n be the degree of the generator polynomial g(x).
1. The Reed-Solomon code can
   - detect up to n errors
   - fill up to n erasures
   - correct up to $\lfloor\frac n2\rfloor$ errors

2. Here
   - "correct errors" means finding error positions together with error values
   - "fill erasures" means finding error values while error positions are given

3. Main implementations
   - detect errors: `haserror(msg, n)`
   - fill erasures: `fill_erases(msg, n)` **(now)**
   - correct errors: `correct_errors(msg, n)` **(next)**
   - fill erasures and correct errors: `correct_msg(msg, n)`

4. Related algorithm
   - find error locator polynomial $\Lambda(x)$: Berlekamp-Massey algorithm
   - find error values via error locator polynomial: Forney algorithm
   - calculate both $\Lambda(x)$ and the error values: Euclidean algorithm(optional)
   - find roots of $\Lambda(x)$: Chien search(we can use brute force search at the beginning)

---

Most of the algorithms are easy to implement, though it may take some effort to understanding.

**Comment by johnnychen94** [comment]
FWIW, we can safely set our Julia lower bound to v1.6 in both `Project.toml` and CI template https://github.com/JuliaImages/QRDecoders.jl/blob/86173f7f8c351ab82fff9fc7da6b46d8d9001c6a/.github/workflows/CI.yml

The overall JuliaImages ecosystem will shift to >= v1.6

**Review by johnnychen94 — APPROVED** [review]
Looks good to me -- no suspicious points. But still need you to check the implementation correctness.

There seem to be a few lines not covered by tests. Might be worth adding some tests for it.

I notice that you've used many `push!`, `pop!` operations. I'm not sure if this will be the performance bottleneck. If so, a possible solution is to pre-allocate the coefficient vector and use StaticArrays to further improve it.

**Comment by RexWzh** [comment]
## main change
1. Add some tests and more functions

    | Function | Description |
    | -------- | ----------- |
    | `derivative_polynomial` | Compute the derivative of a polynomial |
    | `syndrome_decoder` | Forney algorithm, compute error values |
    | `syndrome_decoder!` | as `syndrome_decoder` |
    | `reducebyHorner` | reduce polynomial using Horner's rule(秦九昭算法) |
    | `findroots` | find roots of a polynomial |

2. Add `Manifest.toml` to `docs`

3. Implementation progress
   - find error values via error locator polynomial: Forney algorithm (√)
   - find roots of $\Lambda(x)$ (√) (using Horner's rule with complexity O(N^2) where N ≤ 256)
   - calculate both $\Lambda(x)$ and the error values: Euclidean algorithm(**debugging**)
   - find error locator polynomial $\Lambda(x)$: Berlekamp-Massey algorithm(**current**)

---

The rest of the decoding works requires few algorithms(but might be more tedious).

**Comment by RexWzh** [comment]
> There seem to be a few lines not covered by tests. Might be worth adding some tests for it.

Yes, there'll be more tests after implementing the main algorithms. 

> I notice that you've used many `push!`, `pop!` operations. I'm not sure if this will be the performance bottleneck. If so, a possible solution is to pre-allocate the coefficient vector and use StaticArrays to further improve it.

Got it. Thanks for the helpful advices.

**Comment by RexWzh** [comment]
Implementation of Berlekamp-Massey-algorithm following the discussions [here](https://stackoverflow.com/questions/30215337/errata-erasureserrors-berlekamp-massey-for-reed-solomon-decoding) and the book [Algebraic codes for data transmission](http://qiniu.wzhecnu.cn/Richard%20E.%20Blahut%20-%20Algebraic%20codes%20for%20data%20transmission-Cambridge%20University%20Press%20(2003).pdf).

## Main change
1. rectify words: `fillearsed => fillerased, recieved => received`
2. BM algorithm to find error locator polynomial: `erratalocator_polynomial`
3. BM decoder to recover **message polynomial** from **received polynomial** :`BMdecoder`

Here I use BM algorithm and Forney algorithm to correct errors. The [book](http://qiniu.wzhecnu.cn/Richard%20E.%20Blahut%20-%20Algebraic%20codes%20for%20data%20transmission-Cambridge%20University%20Press%20(2003).pdf) also gives algorithm(figure7.9) on how to do both at the same time. We might consider that later.

Besides, RS-code can decode `v` errors and `ρ` erasures simultaneously with restriction `2 * v + ρ ≤ n`. The erased part(chapter 7.5) is optional and will be added later.

**Comment by RexWzh** [comment]
## implementation plan of the next steps

The coder will be sth looks like this
```jl
mutable struct Decoder
    ## received matrix satisfies some format restrictions
    received_mat::AbstractMatrix
    ## erased positions(optional), note that the indexs start from 0.
    erasures::AbstractVector
    V::Int # version info
    EC::ErrCorrLevel # error correction level
    mask::Int # mask pattern

    ## corrected message matrix, set `nothing` if not called
    corrected_mat::Union{Nothing, AbstractMatrix}
    ## decoded mode: Numeric, Alphanumeric, Byte, Kanji
    mode::Union{Nothing, Mode}
    ## decoded data
    data::Union{Nothing, AbstractString}
    
    function Decoder(mat::AbstractMatrix)
        decoder = new(mat)
        # V =  ## version info
        # fmt = ## format bits
        decoder.EC, decoder.mask = getformat(fmt)
        decoder.corrected_data = nothing
        decode.mode = decode.data = nothing
        decoder
    end
end

function getproperty(decoder::Decoder, key::Symbol)
    v = getfield(decoder, key)
    isnothing(v) || return v
    if key == :corrected_mat
        ...
    elseif key == :mode
        ...
    elseif key == :data
        ...
    end
end
```

**Comment by johnnychen94** [comment]
Unsure of it. The struct definition looks like an over-design for me. -- how would you design the `qrdecode(args...)` API around this?

**Comment by RexWzh** [comment]
> Unsure of it. The struct definition looks like an over-design for me. -- how would you design the `qrdecode(args...)` API around this?

The struct `Decoder` is used to pack the information of the QR code only. Maybe I should write it in this way:
```jl
mutable struct QRInfo
    V::Int # version info(1 ≤ V ≤ 40)
    EC::ErrCorrLevel # error correction level(H, L, M, Q)
    mask::Int # mask pattern(0-7)
    mode::Mode # encoding mode: Numeric, Alphanumeric, Byte, Kanji
    data::AbstractString # decoded data
end

## input: BitArray{2} of the QRCode
function qrdecode(mat::AbstractMatrix; msgonly=true)
    ## check if the matrix is a valid QRCode
    qrcheck(mat) || throw(FormatError()) 

    ## version info
    V = getversion(mat)
    ## format information
    ### error correction level + mask pattern
    fmt = getformat(mat, V)
    EC, mask = qrinfo(mat, fmt)
    
    ## apply mask pattern
    masked_mat = applymask(mat, mask)
    
    ## received data bits(with errors)
    receievdbits = getdatabits(masked_mat)

    ## correct bits(key part -- BMdecoder, Berlekamp Massey algorithm)
    correctbits = RSdecodemsg(receievdbits, V, EC)
    ## sperate into message bits and mode info
    msgbits, mode = getmsgbits(correctbits)
    decodemsg = getdecodemsg(msgbits, mode)
    return msgonly ? decodemsg : QRInfo(V, EC, mask, mode, decodemsg)
end
```

**Comment by johnnychen94** [comment]
I see. It seems that you're mixing the usage of **the interchangeable struct** `QRInfo` and the **decoding status** struct `QRDecodingStatus`. The new `QRInfo` struct is much clearer to me, the other status variables can be local variables created inside the `qrdecode` function.

You can unconditionally return the new `QRInfo` to the callers (for type stability) since they're all useful, I think.

**Comment by RexWzh** [comment]
> I see. It seems that you're mixing the usage of **the interchangeable struct** `QRInfo` and the **decoding status** struct `QRDecodingStatus`. The new `QRInfo` struct is much clearer to me, the other status variables can be local variables created inside the `qrdecode` function.

Yeah, I just notice that. :blush:

Moreover, the variables `V, EC, mode, mask, decodemsg` are necessary to describe a QR-Code, since the same message with different version, mode, error correction level or mask pattern can generate different QR Codes.

> You can unconditionally return the new `QRInfo` to the callers (for type stability) since they're all useful, I think.

Good advice! Got it.

**Comment by RexWzh** [comment]
# Abstract
The current PR is created to implement syndrome decoding algorithm(s) of the RS-code. However, the support of erasures for `BMdecoder` is incomplete(and takes more time to debug), so I replace it by the euclidean decoder.

## Main works
Berlekamp-Massey decoder & Euclidean decoder.

Key functions:
| function | what for | description |
| ---- | ------- | ----------- |
| `erratalocator_polynomial` | calculate the error locator Λx | Berlekamp-Massey algorithm |
| `findroots` | get error positions from Λx | search roots of the given polynomial(by Horner's rule) |
| `forney_algorithm` | compute the error values | - |
| `extended_euclidean_divide` | crucial part of the euclidean decoder | extended euclidean division |
| `euclidean_decoder` | decoder of Reed-Solomon code | - |
| `BMdecoder` | decoder of Reed-Solomon code | - |

## About RS-Code
1. Notation
   - correct an error == find an error position(0-254) and the correspondent error value(0-255)
   - fill an erasure == calculate error value(0-255) given the error position

2. Denote `nmess` the length of the message and `nsym` the number of syndromes s.t. `nmess + nsym ≤ 255`, the resulting RS-Code can:
   - detect up to `nsym` errors
   - correct up to `nsym ÷ 2` errors
   - fill up to `nsym` erasures
   - fill `ρ` erasures and find `v` errors satisfies that `2⋅v + ρ ≤ nsym`

3. received message with too many errors might be uncorrectable or even be misread. (see example below)

### Support for erasures
It is possible for a Reed–Solomon decoder to decode both erasures and errors at the same time, up to a limit (`2⋅v + ρ ≤ nsym`). We first implement the algorithm to support erasures and can consider it later at the application level.

Advantage of using erasures: to increase the error-correction capacity of QR-Code, considering that correcting an error is equivalent to filling two erasures.

Disadvantage: decrease the error-correction capacity of QR-Code if the given positions have no error.


### Example

An example to get a better glance of the RS-code.

1. number of errors within the capacity of the RS-code
    ```jl
    nmess = 1
    nsym = 4
    # 4 bits error correction code + 1 bit message
    # coding rules:
    # [0] =encode> [0, 0, 0, 0, 0]
    # [1] =encode> [64, 120, 54, 15, 1]

    # one bit message `0`
    raw_message = Poly([0])
    encode_message = Poly([0, 0, 0, 0, 0])

    ### transmit ###

    # received message with 2 error
    received_message = Poly([0, 0, 54, 15, 0])
    # v = 2, ρ = 0 and 2 * v + ρ ≤ nsym == 4
    # decoder: I can handle it
    correct_message = Poly([0, 0, 0, 0, 0])
    decode_message = Poly([0])
    ```
    Two errors occurs in the transition, but we can still recover the original message. In fact, it's obvious that `Poly([0, 0, 54, 15, 0])` is closer to `Poly([0, 0, 0, 0, 0])` than `Poly([64, 120, 54, 15, 1])`.

2. number of errors exceed the capacity of the RS-code(detected and error-raised)
   ```jl
   # nmess, nsym, raw_message, decode_message as above
   received_message = Poly([0, 0, 54, 15, 2])
   # v = 3, ρ = 0 and 2 * v + ρ = 6 > sym
   # decoder: raise error
   throw(ReedSolomonError)
   ```
   `Poly([0, 0, 54, 15, 2])` is of the same distance to `Poly([0, 0, 0, 0, 0])` and `Poly([64, 120, 54, 15, 1])`.

3. number of errors exceed the capacity of the RS-code(undetected and misread)
   ```jl
   received_message = Poly([0, 0, 54, 15, 1])
   # reality: v = 3, ρ = 0 and 2 * v + ρ = 6 > sym, raise error
   # decoder: v = 2, ρ = 0 and 2 * v + ρ ≤ nsym == 4, I can handle it
   correct_message = Poly([64, 120, 54, 15, 1])
   decode_message = Poly([1]) # misreading
   ```
   It's true that `Poly([0, 0, 54, 15, 1])` is closer to `Poly([64, 120, 54, 15, 1])` than `Poly([0, 0, 0, 0, 0])`

4. invalid erasures might decrease the capacity of the RS-Code and can cause misreading.
   ```jl
   received_message = Poly([0, 0, 54, 15, 0])
   erasures = [0, 1]
   # received_message = Poly([?, ?, 54, 15, 0])
   # reality: v = 2, ρ = 2 and 2 * v + 2 = 6 > sym, raise error
   # decoder: v = 1, ρ = 2 and 2 * v + 2 = 6 ≤ nsym, I can handle it
   correct_message = Poly([64, 120, 54, 15, 1])
   decode_message = Poly([1]) # misread
   ```
   It's true that `Poly([?, ?, 54, 15, 0])` is closer to `Poly([64, 120, 54, 15, 1])` than `Poly([0, 0, 0, 0, 0])`
   
## Remain works of Syndrome decoding(optional)

1. Add support of erasures for `BMdecoder`
   
2. improve efficiency of the `BMdecoder`(more works to do than Euclidean decoder)

---

I think we can merge this PR for now.

Next steps: [notes about decoding procedures](https://www.wzhecnu.cn/2022/08/20/programming/gsoc/qr-code-06-decode/)

File: src/euclidean.jl [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# Extended Euclidean division algorithm
+using QRDecoders: ReedSolomonError
```
Comment (johnnychen94) [inline src/euclidean.jl]: Actually not needed.

**Review by johnnychen94 — APPROVED** [review]

**Pull request merged**